### PR TITLE
refactor(runtime): tighten orchestration payload boundaries

### PR DESCRIPTION
**Pull request by Maxsky5**
## Summary
- tighten session, task metadata, and runtime failure payload parsing so invalid runtime-orchestration data fails fast at the intended boundary
- preserve durable session compatibility and existing batching behavior while removing unsafe hot-path casts and broad record coercions
- add regression coverage for recursive permission metadata, legacy metadata normalization, indexed persisted-session errors, malformed token/source payloads, and mixed schema-error handling

## Goal
This PR hardens the runtime and session typing boundaries used by the orchestration flow. The goal is to stop malformed payloads at the nearest contract or adapter edge instead of letting loosely typed data leak deeper into the system, while keeping the durable session model stable and avoiding fallback behavior.

## Context
The task focused on boundary safety across the shared contracts, the Tauri host adapter, the OpenCode SDK adapter, and the desktop session event batcher. These paths sit between persisted task/session metadata, runtime ensure failures, streamed message payloads, and the frontend orchestration state. They already had compatibility expectations around durable session reads, legacy metadata normalization, and batching semantics, so the change needed to tighten validation without widening the persisted contract or changing behavior that downstream tests rely on.

## Technical Choices
- tightened `agentSessionPermissionRequestSchema.metadata` to a bounded recursive value schema instead of `unknown`, while keeping `agentSessionRecordSchema` durable-only and non-strict for legacy persisted fields
- tightened `taskMetadataPayloadSchema.agentSessions` to canonical session records and kept legacy delivery normalization, then preserved actionable persisted-session errors by remapping all-session validation failures into the existing indexed `agentSessions[i]...` format in the Tauri metadata cache
- switched metadata normalization to field-by-field precedence so a present top-level `pullRequest` no longer blocks legacy `directMerge` normalization, and vice versa
- replaced broad record casts in the OpenCode SDK message normalizers with explicit guarded parsing for source text and token breakdown payloads so malformed nested values are ignored safely instead of accessed unchecked
- replaced the runtime ensure adapter's loose failure probing with a typed failure-envelope parser that preserves the existing precedence for `failureKind`, `message`, nested `cause`, and outer error handling
- removed `as never` hot-path casts from the desktop session event batcher through an exhaustive typed dispatch over `SessionEvent[\"type\"]`, while keeping supersession, dedupe, merge, and emit timing behavior unchanged
- added regression tests across contracts, adapters, and desktop orchestration to lock in the compatibility and fail-fast expectations above

## Verification
- `bun run typecheck`
- `bun run lint`
- `bun run test`
- `bun run build`
- `bun run check:rust`
- `bun run test:rust`
- `cargo fmt --all --check`
- `cargo clippy --workspace --all-targets -- -D warnings`

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved resilience when processing malformed message data and metadata; the system now handles invalid formats gracefully without data loss.
  * Enhanced error messages for metadata validation to better pinpoint issues.

* **Improvements**
  * Refined message display styling.
  * Strengthened validation for agent session metadata.

* **Tests**
  * Added test coverage for edge cases with malformed payloads and invalid data structures.

### CODE REVIEW

File: packages/adapters-opencode-sdk/src/message-normalizers.ts [hunk 1]
```diff
@@ -51,14 +51,14 @@ const readFilePathFromUrl = (url: string): string | null => {
...
     return undefined;
   }
 
-  const record = value as Record<string, unknown>;
-  const textValue = record.value;
-  const start = record.start;
-  const end = record.end;
+  const textValue = readUnknownProp(record, "value");
+  const start = readNumberProp(record, ["start"]);
+  const end = readNumberProp(record, ["end"]);
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/adapters-opencode-sdk/src/message-normalizers.ts:61]: Keeping this as-is. `readNumberProp` is intentionally the path-aware reader in this adapter layer and already accepts key arrays across the package, so the single-element array form is part of the current helper API rather than an accidental inconsistency. I did not want to mix a local call-site cleanup with a broader helper-shape change in this PR.

File: packages/adapters-opencode-sdk/src/message-normalizers.ts [hunk 2]
```diff
@@ -51,14 +51,14 @@ const readFilePathFromUrl = (url: string): string | null => {
...
   if (typeof textValue !== "string" || typeof start !== "number" || typeof end !== "number") {
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/adapters-opencode-sdk/src/message-normalizers.ts:62]: Leaving this unchanged. In this path `readNumberProp(...)` already rejects non-finite numeric values, so by the time `start` and `end` reach this check they are either finite numbers or `undefined`. Adding `Number.isFinite(...)` here would duplicate the guard without tightening behavior further.

File: packages/adapters-opencode-sdk/src/message-normalizers.ts [hunk 3]
```diff
@@ -343,6 +343,34 @@ const toFiniteNumber = (value: unknown): number | null => {
...
+  if (!record) {
+    return undefined;
+  }
+
+  const input = readNumberProp(record, ["input"]);
+  const output = readNumberProp(record, ["output"]);
+  const reasoning = readNumberProp(record, ["reasoning"]);
+  const cacheRecord = readRecordProp(record, "cache");
+  const cacheRead = cacheRecord ? readNumberProp(cacheRecord, ["read"]) : undefined;
+  const cacheWrite = cacheRecord ? readNumberProp(cacheRecord, ["write"]) : undefined;
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/adapters-opencode-sdk/src/message-normalizers.ts:357]: Keeping the current form for the same reason as above: `readNumberProp` is the path-aware reader in this package and its array-based call shape is already used consistently for numeric reads. I did not want to partially redesign that helper API inside this boundary-tightening PR.

File: packages/adapters-tauri-host/src/task-metadata-cache.ts [hunk 4]
```diff
@@ -1,11 +1,15 @@
...
-  agentSessionRecordSchema,
   type TaskMetadataPayload,
   taskMetadataPayloadSchema,
 } from "@openducktor/contracts";
 import type { InvokeFn } from "./invoke-utils";
 
+type MetadataIssue = {
+  path: PropertyKey[];
+  message: string;
+};
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/adapters-tauri-host/src/task-metadata-cache.ts:7]: Applied in `3a4b3d9c`. `MetadataIssue.path` is now narrowed to `Array<string | number>`, with a small conversion step from the raw Zod issue shape so we keep the local type precise without fighting the upstream bundled type surface.

File: packages/adapters-tauri-host/src/task-metadata-cache.ts [hunk 5]
```diff
@@ -1,11 +1,15 @@
...
+
 export type ParsedTaskMetadata = Omit<TaskMetadataPayload, "agentSessions"> & {
   agentSessions: AgentSessionRecord[];
 };
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/adapters-tauri-host/src/task-metadata-cache.ts]: Applied in `3a4b3d9c`. `ParsedTaskMetadata` now aliases `TaskMetadataPayload` directly because the schema already guarantees canonical `agentSessions` typing, so the previous override type had become redundant.

File: packages/contracts/src/session-schemas.ts [hunk 6]
```diff
@@ -14,6 +14,20 @@ export type AgentSessionScenario = z.infer<typeof agentSessionScenarioSchema>;
...
+> = z.lazy(() =>
+  z.union([
+    z.string(),
+    z.number(),
+    z.boolean(),
+    z.null(),
+    z.undefined(),
+    z.array(agentSessionMetadataValueSchema),
+    z.record(z.string(), agentSessionMetadataValueSchema),
+  ]),
```
[the remark this answers is not in the document]
Comment (Maxsky5) [reply packages/contracts/src/session-schemas.ts:28]: I intentionally left `z.undefined()` in place for now. Tightening this to JSON-only values would be a behavioral boundary change rather than a pure cleanup, and I do not have explicit confirmation from the existing persisted/session payload surfaces that `undefined` metadata is disallowed. I kept the current compatibility shape in this PR rather than silently narrowing it.
